File: SEARCH/interpolationSearch/interpolationSearch.cpp
#include <iostream>
using namespace std;

// Функция реализует интерполяционный поиск в отсортированном массиве
int interpolationSearch(int arr[], int n, int x) {
    int low = 0;              // Нижний индекс области поиска
    int high = n - 1;         // Верхний индекс области поиска

    while ((low <= high) && (x >= arr[low]) && (x <= arr[high])) {
        // Вычисление следующей позиции поиска
        int pos = low + (((double)(high - low) / (arr[high] - arr[low])) * (x - arr[low]));

        // Найденный элемент?
        if (arr[pos] == x) {
            return pos;       // Вернуть индекс найденного элемента
        }

        // Продолжаем поиск слева, если элемент меньше текущего
        if (arr[pos] < x) {
            low = pos + 1;
        }
        // Или справа, если больше
        else {
            high = pos - 1;
        }
    }

    return -1;               // Если элемент не найден, вернуть -1
}

// Основная программа для демонстрации работы поиска
int main() {
    int arr[] = {10, 12, 13, 16, 18, 19, 20, 21, 22, 23};
    int n = sizeof(arr)/sizeof(arr[0]);
    int x = 18;                   // Искать элемент '18'

    int index = interpolationSearch(arr, n, x);

    if(index != -1) {
        cout << "Элемент найден на индексе: " << index << endl;
    } else {
        cout << "Элемент не найден." << endl;
    }

    return 0;
}

File: lab6/SEARCH/interpolationSearch/interpolationSearch.java
public class interpolationSearch {

    // Реализация интерполяционного поиска
    public static int interpolationSearch(int[] arr, int x) {
        int low = 0;                     // Нижняя граница поиска
        int high = arr.length - 1;       // Верхняя граница поиска

        while ((low <= high) && (x >= arr[low]) && (x <= arr[high])) {
            // Оцениваем потенциальную позицию искомого элемента
            double pos = low + ((double)(high - low) / (arr[high] - arr[low])) * (x - arr[low]);
            
            // Преобразование в целое число (так как индекс массива должен быть целым числом)
            int mid = (int)Math.floor(pos);

            // Нашли элемент!
            if (arr[mid] == x) {
                return mid;              // Индекс найденного элемента
            }

            // Если элемент меньше среднего, продолжаем поиск слева
            if (arr[mid] < x) {
                low = mid + 1;
            }
            // Иначе — справа
            else {
                high = mid - 1;
            }
        }

        return -1;                      // Если элемент не найден
    }

    // Точка входа в программу
    public static void main(String[] args) {
        int[] sortedArray = {10, 12, 13, 16, 18, 19, 20, 21, 22, 23};
        int elementToFind = 18;          // Значение, которое мы хотим найти

        // Выполняем поиск
        int index = interpolationSearch(sortedArray, elementToFind);

        // Результат
        if (index != -1) {
            System.out.println("Элемент найден на индексе: " + index);
        } else {
            System.out.println("Элемент не найден.");
        }
    }
}